File: gameUI.js
/**
 * UI components and rendering for the game.
 */
class GameUI {
    constructor() {
        // UI dimensions
        this.HEALTH_BAR_HEIGHT = 15;
        this.LEVEL_BAR_HEIGHT = 8;
    }
    
    /**
     * Draw the game background.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {number} width The width of the game area
     * @param {number} height The height of the game area
     */
    drawBackground(ctx, width, height) {
        ctx.fillStyle = GameColors.BACKGROUND;
        ctx.fillRect(0, 0, width, height);
        
        // Add grid lines for retro effect
        ctx.strokeStyle = '#222222';
        ctx.lineWidth = 1;
        
        for (let i = 0; i < width; i += 20) {
            ctx.beginPath();
            ctx.moveTo(i, 0);
            ctx.lineTo(i, height);
            ctx.stroke();
        }
        
        for (let i = 0; i < height; i += 20) {
            ctx.beginPath();
            ctx.moveTo(0, i);
            ctx.lineTo(width, i);
            ctx.stroke();
        }
    }
    
    /**
     * Draw the player health bar.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {number} currentHealth Current health points
     * @param {number} maxHealth Maximum health points
     * @param {number} x X position of the bar
     * @param {number} y Y position of the bar
     * @param {number} width Width of the bar
     */
    drawHealthBar(ctx, currentHealth, maxHealth, x, y, width) {
        // Draw health background
        ctx.fillStyle = '#333333';
        ctx.fillRect(x, y, width, this.HEALTH_BAR_HEIGHT);
        
        // Calculate health width
        let healthWidth = Math.floor(width * Math.max(0, currentHealth / maxHealth));
        
        // Choose color based on health percentage
        let healthPercent = currentHealth / maxHealth;
        if (healthPercent > 0.6) {
            ctx.fillStyle = GameColors.SUCCESS_GREEN;
        } else if (healthPercent > 0.3) {
            ctx.fillStyle = GameColors.ACCENT_YELLOW;
        } else {
            ctx.fillStyle = GameColors.HEALTH_RED;
        }
        
        // Draw health bar
        ctx.fillRect(x, y, healthWidth, this.HEALTH_BAR_HEIGHT);
        
        // Draw border
        ctx.strokeStyle = GameColors.TEXT;
        ctx.lineWidth = 1;
        ctx.strokeRect(x, y, width, this.HEALTH_BAR_HEIGHT);
        
        // Draw health text
        ctx.fillStyle = GameColors.TEXT;
        ctx.font = '12px monospace';
        let healthText = currentHealth + '/' + maxHealth;
        let textWidth = ctx.measureText(healthText).width;
        ctx.fillText(healthText, x + (width - textWidth) / 2, y + 12);
    }
    
    /**
     * Draw the level progress bar.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {Level} level The current level
     * @param {number} x X position of the bar
     * @param {number} y Y position of the bar
     * @param {number} width Width of the bar
     */
    drawLevelProgressBar(ctx, level, x, y, width) {
        // Draw level progress background
        ctx.fillStyle = '#333333';
        ctx.fillRect(x, y, width, this.LEVEL_BAR_HEIGHT);
        
        // Calculate progress width
        let progressWidth = Math.floor(width * level.getProgressPercentage());
        
        // Draw progress bar
        ctx.fillStyle = GameColors.ACCENT_YELLOW;
        ctx.fillRect(x, y, progressWidth, this.LEVEL_BAR_HEIGHT);
        
        // Draw border
        ctx.strokeStyle = GameColors.TEXT;
        ctx.lineWidth = 1;
        ctx.strokeRect(x, y, width, this.LEVEL_BAR_HEIGHT);
    }
    
    /**
     * Draw game stats.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {Level} level Current level
     * @param {number} score Current score
     * @param {number} lives Remaining lives
     * @param {string} playerShape Current player shape
     */
    drawGameStats(ctx, level, score, lives, playerShape) {
        ctx.fillStyle = GameColors.TEXT;
        ctx.font = 'bold 14px monospace';
        
        // Use consistent spacing for all items in the HUD
        let y = 25;
        let lineHeight = 22;
        
        // Level indicator
        ctx.fillText('LEVEL: ' + level.levelNumber, 10, y);
        y += lineHeight;
        
        // Score
        ctx.fillText('SCORE: ' + score, 10, y);
        y += lineHeight;
        
        // Lives with heart symbols
        ctx.fillText('LIVES: ', 10, y);
        for (let i = 0; i < lives; i++) {
            // Draw simple pixel heart
            ctx.fillStyle = GameColors.HEALTH_RED;
            ctx.fillRect(70 + i * 20, y - 10, 5, 5);
            ctx.fillRect(75 + i * 20, y - 10, 5, 5);
            ctx.fillRect(65 + i * 20, y - 5, 15, 5);
            ctx.fillRect(70 + i * 20, y, 5, 5);
            ctx.fillStyle = GameColors.TEXT;
        }
        y += lineHeight;
        
        // Current shape
        ctx.fillText('SHAPE: ' + playerShape, 10, y);
        y += lineHeight;
        
        // Controls reminder
        ctx.font = '12px monospace';
        ctx.fillText('1-2-3: Change Shape   |   ARROWS: Move   |   SPACE: Fire', 10, y + 15);
    }
    
    /**
     * Draw enemy health bar.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {object} enemyBounds The enemy bounds {x, y, width, height}
     * @param {number} health Current health
     * @param {number} maxHealth Maximum health
     * @param {boolean} matchingPlayerShape Whether the player shape matches the enemy
     */
    drawEnemyHealthBar(ctx, enemyBounds, health, maxHealth, matchingPlayerShape) {
        // Draw health background
        ctx.fillStyle = '#333333';
        ctx.fillRect(enemyBounds.x, enemyBounds.y - 10, enemyBounds.width, 5);
        
        // Draw health bar
        ctx.fillStyle = GameColors.HEALTH_RED;
        let healthWidth = Math.floor(enemyBounds.width * (health / maxHealth));
        ctx.fillRect(enemyBounds.x, enemyBounds.y - 10, healthWidth, 5);
        
        // Highlight enemies matching player shape for critical hits
        if (matchingPlayerShape) {
            ctx.fillStyle = this.hexToRgba(GameColors.ACCENT_YELLOW, 0.4);
            ctx.beginPath();
            ctx.arc(
                enemyBounds.x + enemyBounds.width / 2,
                enemyBounds.y + enemyBounds.height / 2,
                enemyBounds.width / 2 + 5,
                0,
                Math.PI * 2
            );
            ctx.fill();
        }
    }
    
    /**
     * Draw game over screen.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {number} width The width of the game area
     * @param {number} height The height of the game area
     * @param {number} finalScore The final score
     * @param {number} highestLevel The highest level reached
     */
    drawGameOverScreen(ctx, width, height, finalScore, highestLevel) {
        // Semi-transparent overlay
        ctx.fillStyle = 'rgba(0, 0, 0, 0.8)';
        ctx.fillRect(0, 0, width, height);
        
        ctx.font = 'bold 30px monospace';
        ctx.fillStyle = GameColors.HEALTH_RED;
        
        let gameOver = 'GAME OVER';
        let textWidth = ctx.measureText(gameOver).width;
        ctx.fillText(gameOver, (width - textWidth) / 2, height / 2 - 30);
        
        ctx.font = 'bold 14px monospace';
        ctx.fillStyle = GameColors.TEXT;
        
        let scoreText = 'FINAL SCORE: ' + finalScore;
        let levelText = 'HIGHEST LEVEL: ' + highestLevel;
        
        textWidth = ctx.measureText(scoreText).width;
        ctx.fillText(scoreText, (width - textWidth) / 2, height / 2 + 10);
        
        textWidth = ctx.measureText(levelText).width;
        ctx.fillText(levelText, (width - textWidth) / 2, height / 2 + 40);
        
        let restartText = 'PRESS ENTER TO RESTART';
        ctx.font = '12px monospace';
        ctx.fillStyle = GameColors.ACCENT_YELLOW;
        textWidth = ctx.measureText(restartText).width;
        ctx.fillText(restartText, (width - textWidth) / 2, height / 2 + 80);
    }
    
    /**
     * Draw the start screen.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {number} width The width of the game area
     * @param {number} height The height of the game area
     * @param {number} timestamp Current timestamp for animations
     */
    drawStartScreen(ctx, width, height, timestamp) {
        // Title
        ctx.font = 'bold 36px monospace';
        ctx.fillStyle = GameColors.ACCENT_YELLOW;
        let title = 'SHAPE SHIFTER DUEL';
        let textWidth = ctx.measureText(title).width;
        ctx.fillText(title, (width - textWidth) / 2, height / 3);
        
        // Subtitle
        ctx.font = 'bold 14px monospace';
        ctx.fillStyle = GameColors.TEXT;
        let subtitle = 'LEVEL-BASED ARCADE SHOOTER';
        textWidth = ctx.measureText(subtitle).width;
        ctx.fillText(subtitle, (width - textWidth) / 2, height / 3 + 40);
        
        // Instructions
        ctx.font = '12px monospace';
        let instructions = [
            'CONTROLS:',
            'ARROWS: Move',
            'SPACE: Fire',
            '1-2-3: Change Shape',
            'P: Pause Game'
        ];
        
        let y = height / 2;
        for (let instruction of instructions) {
            textWidth = ctx.measureText(instruction).width;
            ctx.fillText(instruction, (width - textWidth) / 2, y);
            y += 25;
        }
        
        // Start prompt - blinking effect
        if (Math.floor(timestamp / 500) % 2 === 0) {
            ctx.font = 'bold 14px monospace';
            ctx.fillStyle = GameColors.SUCCESS_GREEN;
            let startPrompt = 'PRESS ENTER TO START';
            textWidth = ctx.measureText(startPrompt).width;
            ctx.fillText(startPrompt, (width - textWidth) / 2, height * 3 / 4 + 40);
        }
    }
    
    /**
     * Convert hex color to rgba.
     * 
     * @param {string} hex The hex color
     * @param {number} alpha The alpha value
     * @return {string} The rgba color string
     */
    hexToRgba(hex, alpha) {
        let r = parseInt(hex.slice(1, 3), 16);
        let g = parseInt(hex.slice(3, 5), 16);
        let b = parseInt(hex.slice(5, 7), 16);
        return `rgba(${r}, ${g}, ${b}, ${alpha})`;
    }
}

File: GameUI.java
import javax.swing.*;
import java.awt.*;

/**
 * UI components and rendering for the game.
 */
public class GameUI {
    private Font pixelFont;
    private Font smallPixelFont;
    
    // UI dimensions
    private final int HEALTH_BAR_HEIGHT = 15;
    private final int LEVEL_BAR_HEIGHT = 8;
    
    public GameUI() {
        // Try to load a pixel font, fallback to monospaced
        try {
            // Use Monospaced as our "pixel" font as it's available on all systems
            pixelFont = new Font("Monospaced", Font.BOLD, 14);
            smallPixelFont = new Font("Monospaced", Font.PLAIN, 12);
        } catch (Exception e) {
            // Fallback
            pixelFont = new Font("SansSerif", Font.BOLD, 14);
            smallPixelFont = new Font("SansSerif", Font.PLAIN, 12);
        }
    }
    
    /**
     * Draw the game background.
     * 
     * @param g The graphics context
     * @param width The width of the game area
     * @param height The height of the game area
     */
    public void drawBackground(Graphics g, int width, int height) {
        g.setColor(GameColors.BACKGROUND);
        g.fillRect(0, 0, width, height);
        
        // Add grid lines for retro effect
        g.setColor(new Color(0x222222));
        for (int i = 0; i < width; i += 20) {
            g.drawLine(i, 0, i, height);
        }
        for (int i = 0; i < height; i += 20) {
            g.drawLine(0, i, width, i);
        }
    }
    
    /**
     * Draw the player health bar.
     * 
     * @param g The graphics context
     * @param currentHealth Current health points
     * @param maxHealth Maximum health points
     * @param x X position of the bar
     * @param y Y position of the bar
     * @param width Width of the bar
     */
    public void drawHealthBar(Graphics g, int currentHealth, int maxHealth, int x, int y, int width) {
        // Draw health background
        g.setColor(new Color(0x333333));
        g.fillRect(x, y, width, HEALTH_BAR_HEIGHT);
        
        // Calculate health width
        int healthWidth = (int)(width * Math.max(0, (double)currentHealth / maxHealth));
        
        // Choose color based on health percentage
        double healthPercent = (double)currentHealth / maxHealth;
        if (healthPercent > 0.6) {
            g.setColor(GameColors.SUCCESS_GREEN);
        } else if (healthPercent > 0.3) {
            g.setColor(GameColors.ACCENT_YELLOW);
        } else {
            g.setColor(GameColors.HEALTH_RED);
        }
        
        // Draw health bar
        g.fillRect(x, y, healthWidth, HEALTH_BAR_HEIGHT);
        
        // Draw border
        g.setColor(GameColors.TEXT);
        g.drawRect(x, y, width, HEALTH_BAR_HEIGHT);
        
        // Draw health text
        g.setFont(smallPixelFont);
        String healthText = currentHealth + "/" + maxHealth;
        FontMetrics fm = g.getFontMetrics();
        int textX = x + (width - fm.stringWidth(healthText)) / 2;
        int textY = y + fm.getAscent() + (HEALTH_BAR_HEIGHT - fm.getHeight()) / 2;
        g.drawString(healthText, textX, textY);
    }
    
    /**
     * Draw the level progress bar.
     * 
     * @param g The graphics context
     * @param level The current level
     * @param x X position of the bar
     * @param y Y position of the bar
     * @param width Width of the bar
     */
    public void drawLevelProgressBar(Graphics g, Level level, int x, int y, int width) {
        // Draw level progress background
        g.setColor(new Color(0x333333));
        g.fillRect(x, y, width, LEVEL_BAR_HEIGHT);
        
        // Calculate progress width
        int progressWidth = (int)(width * level.getProgressPercentage());
        
        // Draw progress bar
        g.setColor(GameColors.ACCENT_YELLOW);
        g.fillRect(x, y, progressWidth, LEVEL_BAR_HEIGHT);
        
        // Draw border
        g.setColor(GameColors.TEXT);
        g.drawRect(x, y, width, LEVEL_BAR_HEIGHT);
    }
    
    /**
     * Draw game stats.
     * 
     * @param g The graphics context
     * @param level Current level
     * @param score Current score
     * @param lives Remaining lives
     * @param playerShape Current player shape
     */
    public void drawGameStats(Graphics g, Level level, int score, int lives, String playerShape) {
        g.setColor(GameColors.TEXT);
        g.setFont(pixelFont);
        
        // Use consistent spacing for all items in the HUD
        int y = 25;
        int lineHeight = 22;
        
        // Level indicator
        g.drawString("LEVEL: " + level.getLevelNumber(), 10, y);
        y += lineHeight;
        
        // Score
        g.drawString("SCORE: " + score, 10, y);
        y += lineHeight;
        
        // Lives with heart symbols
        g.drawString("LIVES: ", 10, y);
        for (int i = 0; i < lives; i++) {
            // Draw simple pixel heart
            g.setColor(GameColors.HEALTH_RED);
            g.fillRect(70 + i * 20, y - 10, 5, 5);
            g.fillRect(75 + i * 20, y - 10, 5, 5);
            g.fillRect(65 + i * 20, y - 5, 15, 5);
            g.fillRect(70 + i * 20, y, 5, 5);
            g.setColor(GameColors.TEXT);
        }
        y += lineHeight;
        
        // Current shape
        g.drawString("SHAPE: " + playerShape, 10, y);
        y += lineHeight;
        
        // Controls reminder
        g.setFont(smallPixelFont);
        g.drawString("1-2-3: Change Shape   |   ARROWS: Move   |   SPACE: Fire", 10, y + 15);
    }
    
    /**
     * Draw enemy health bar.
     * 
     * @param g The graphics context
     * @param enemy The enemy object
     * @param health Current health
     * @param maxHealth Maximum health
     * @param matchingPlayerShape Whether the player shape matches the enemy
     */
    public void drawEnemyHealthBar(Graphics g, Rectangle enemyBounds, int health, int maxHealth, boolean matchingPlayerShape) {
        // Draw health background
        g.setColor(new Color(0x333333));
        g.fillRect(enemyBounds.x, enemyBounds.y - 10, enemyBounds.width, 5);
        
        // Draw health bar
        g.setColor(GameColors.HEALTH_RED);
        int healthWidth = (int)(enemyBounds.width * ((double)health / maxHealth));
        g.fillRect(enemyBounds.x, enemyBounds.y - 10, healthWidth, 5);
        
        // Highlight enemies matching player shape for critical hits
        if (matchingPlayerShape) {
            g.setColor(new Color(GameColors.ACCENT_YELLOW.getRed(), 
                              GameColors.ACCENT_YELLOW.getGreen(), 
                              GameColors.ACCENT_YELLOW.getBlue(), 100));
            g.fillOval(enemyBounds.x - 5, enemyBounds.y - 5, 
                       enemyBounds.width + 10, enemyBounds.height + 10);
        }
    }
    
    /**
     * Draw game over screen.
     * 
     * @param g The graphics context
     * @param width The width of the game area
     * @param height The height of the game area
     * @param finalScore The final score
     * @param highestLevel The highest level reached
     */
    public void drawGameOverScreen(Graphics g, int width, int height, int finalScore, int highestLevel) {
        // Semi-transparent overlay
        g.setColor(new Color(0, 0, 0, 200));
        g.fillRect(0, 0, width, height);
        
        g.setFont(new Font("Monospaced", Font.BOLD, 30));
        g.setColor(GameColors.HEALTH_RED);
        
        String gameOver = "GAME OVER";
        FontMetrics fmBig = g.getFontMetrics();
        int textX = (width - fmBig.stringWidth(gameOver)) / 2;
        g.drawString(gameOver, textX, height / 2 - 30);
        
        g.setFont(pixelFont);
        g.setColor(GameColors.TEXT);
        
        String scoreText = "FINAL SCORE: " + finalScore;
        String levelText = "HIGHEST LEVEL: " + highestLevel;
        
        FontMetrics fm = g.getFontMetrics();
        
        g.drawString(scoreText, (width - fm.stringWidth(scoreText)) / 2, height / 2 + 10);
        g.drawString(levelText, (width - fm.stringWidth(levelText)) / 2, height / 2 + 40);
        
        String restartText = "PRESS ENTER TO RESTART";
        g.setFont(smallPixelFont);
        g.setColor(GameColors.ACCENT_YELLOW);
        g.drawString(restartText, 
                  (width - g.getFontMetrics().stringWidth(restartText)) / 2, 
                  height / 2 + 80);
    }
    
    /**
     * Get the pixel font.
     * 
     * @return The pixel font
     */
    public Font getPixelFont() {
        return pixelFont;
    }
    
    /**
     * Get the small pixel font.
     * 
     * @return The small pixel font
     */
    public Font getSmallPixelFont() {
        return smallPixelFont;
    }
}
